File: MultiCraft.Unity/Assets/Multicraft/Scripts/Engine/Core/Worlds/WorldGenerator.cs
using System.Collections.Generic;
using MultiCraft.Scripts.Engine.Core.Biomes;
using UnityEngine;

namespace MultiCraft.Scripts.Engine.Core.Worlds
{
    [CreateAssetMenu(fileName = "WorldGenerator", menuName = "MultiCraft/WorldGenerator")]
    public class WorldGenerator : ScriptableObject
    {
        [Header("Settings")] public int BaseHeight = 64;
        public int WaterLevel = 63;
        public float RiverChance = 0f;
        public float TreeFrequency = 0.5f;

        [Header("Biomes")] public Biome[] Biomes;
        private Biome[] _biomes;
        private float _biomesChance = -1f;

        [Header("Noises Settings")] public NoiseOctaveSetting BiomeNoiseOctaves;
        private FastNoiseLite _biomeNoise;


        public NoiseOctaveSetting FloraNoiseOctaves;
        private FastNoiseLite _floraNoise;

        public NoiseOctaveSetting[] TreeNoiseOctaves;
        private FastNoiseLite[] _treeNoise;


        public NoiseOctaveSetting WaterNoiseOctaves;
        private FastNoiseLite _waterNoise;

        public NoiseOctaveSetting[] SurfaceNoiseOctaves;
        private FastNoiseLite[] _surfaceNoise;


        public void Initialize(int seed)
        {
            InitializeNoises(seed);
            InitializeBiomes();
        }

        private void InitializeBiomes()
        {
            _biomesChance = 0f;
            _biomes = new Biome[Biomes.Length];
            for (int i = 0; i < _biomes.Length; i++)
            {
                _biomes[i] = new Biome(Biomes[i]);
                _biomesChance += _biomes[i].biomeChance;
                _biomes[i].biomeChance = _biomesChance;
            }

            for (int i = 0; i < _biomes.Length; i++)
            {
                _biomes[i].biomeChance /= _biomesChance;
            }
        }

        public void InitializeNoises(int seed)
        {
            _surfaceNoise = new FastNoiseLite[SurfaceNoiseOctaves.Length];
            for (var i = 0; i < SurfaceNoiseOctaves.Length; i++)
            {
                _surfaceNoise[i] = InitializeNoise(SurfaceNoiseOctaves[i], seed);
            }

            _biomeNoise = InitializeNoise(BiomeNoiseOctaves, seed);
            _waterNoise = InitializeNoise(WaterNoiseOctaves, seed);
            _floraNoise = InitializeNoise(FloraNoiseOctaves, seed);

            _treeNoise = new FastNoiseLite[TreeNoiseOctaves.Length];
            for (var i = 0; i < TreeNoiseOctaves.Length; i++)
            {
                _treeNoise[i] = InitializeNoise(TreeNoiseOctaves[i], seed);
            }
        }

        private FastNoiseLite InitializeNoise(NoiseOctaveSetting setting, int seed)
        {
            FastNoiseLite noise = new FastNoiseLite(seed);
            noise.SetNoiseType(setting.NoiseType);
            noise.SetFrequency(setting.Frequency);

            noise.SetFractalType(setting.FractalType);
            noise.SetFractalOctaves(setting.FractalOctaves);
            noise.SetFractalGain(setting.FractalGain);

            noise.SetCellularDistanceFunction(setting.CellularDistanceFunction);
            noise.SetCellularReturnType(setting.CellularReturnType);
            noise.SetCellularJitter(setting.CellularJitter);

            noise.SetDomainWarpType(setting.DomainWarpType);
            noise.SetDomainWarpAmp(setting.DomainWarpAmplitude);
            return noise;
        }

        public int[,,] Generate(int xOffset, int yOffset, int zOffset, ref int[,] surface)
        {
            var blocks = new int[World.ChunkWidth, World.ChunkHeight, World.ChunkWidth];

            GenerateSurface(ref blocks, out var surfaceHeight, xOffset, yOffset, zOffset);
            GenerateTrees(ref blocks, surfaceHeight, xOffset, yOffset, zOffset);
            surface = surfaceHeight;

            return blocks;
        }

        private void GenerateTrees(ref int[,,] blocks, int[,] surfaceHeight, int xOffset, int yOffset, int zOffset)
        {
            Biome biome = new Biome(_biomes[0]);

            for (int x = 0; x < World.ChunkWidth; x++)
            {
                for (int z = 0; z < World.ChunkWidth; z++)
                {
                    int height = surfaceHeight[x, z];

                    float treeNoiseValue = GetTreeNoise(x + xOffset, z + zOffset);

                    if (treeNoiseValue > TreeFrequency && height > BaseHeight - 1 && height != -1)
                    {
                        var biomeType = (GetBiomeNoise(x + xOffset, z + zOffset) + 1f) / 2f;
                        for (int b = 0; b < _biomes.Length; b++)
                        {
                            if (biomeType <= _biomes[b].biomeChance)
                            {
                                biome = new Biome(_biomes[b]);
                                break;
                            }
                        }

                        var treeType = Random.Range(0, biome.treeType.Length);
                        var leavesType = treeType;

                        if (biome.treeType.Length == 0) break;

                        int treeHeight = Random.Range(biome.minTreeHeight[treeType], biome.maxTreeHeight[treeType] + 1);
                        int treeChance = Random.Range(1, 100);
                        int foliageRadius;
                        if (biome.treeType[treeType] != 99)
                        {
                            foliageRadius = 3;
                            if (x + foliageRadius is >= 0 and < World.ChunkWidth &&
                                x - foliageRadius is >= 0 and < World.ChunkWidth &&
                                z + foliageRadius is >= 0 and < World.ChunkWidth &&
                                z - foliageRadius is >= 0 and < World.ChunkWidth &&
                                blocks[x, height - 1, z] != 0 &&
                                treeChance <= biome.treeChance[treeType])
                            {
                                int foliageY = height + treeHeight - 1;
                                for (int dy = 0; dy <= foliageRadius; dy++)
                                {
                                    for (int dx = -foliageRadius; dx <= foliageRadius; dx++)
                                    {
                                        for (int dz = -foliageRadius; dz <= foliageRadius; dz++)
                                        {
                                            int distance = dx * dx + dy * dy + dz * dz;
                                            if (distance <= foliageRadius * foliageRadius - 1 &&
                                                foliageY < World.ChunkHeight)
                                            {
                                                if (biome.leavesType[leavesType] != -1)
                                                    blocks[x + dx, foliageY + dy, z + dz] =
                                                        biome.leavesType[leavesType];
                                            }
                                        }
                                    }
                                }

                                for (int dy = 1; dy <= treeHeight; dy++)
                                    blocks[x, height + dy, z] = biome.treeType[treeType];
                            }
                        }
                        else
                        {
                            foliageRadius = 5;
                            if (x + foliageRadius is >= 0 and < World.ChunkWidth &&
                                x - foliageRadius is >= 0 and < World.ChunkWidth &&
                                z + foliageRadius is >= 0 and < World.ChunkWidth &&
                                z - foliageRadius is >= 0 and < World.ChunkWidth &&
                                blocks[x, height - 1, z] != 0 &&
                                treeChance <= biome.treeChance[treeType])
                            {
                                int oddFoliageRadius = 0;
                                int evenFoliageRadius = 0;
                                int ddy = 0;
                                for (int dy = height + treeHeight - 1;
                                     dy >= height + 1 + Random.Range(1, 3);
                                     ddy++, dy -= 1)
                                {
                                    if (ddy % 2 == 1)
                                    {
                                        oddFoliageRadius += 1;
                                        foliageRadius = Mathf.Clamp(oddFoliageRadius, 1, 4);
                                    }
                                    else
                                    {
                                        evenFoliageRadius += 1;
                                        foliageRadius = Mathf.Clamp(evenFoliageRadius + 1, 1, 5);
                                    }

                                    for (int dx = -foliageRadius; dx <= foliageRadius; dx++)
                                    {
                                        for (int dz = -foliageRadius; dz <= foliageRadius; dz++)
                                        {
                                            int distance = dx * dx + dz * dz;
                                            if (distance <= foliageRadius * foliageRadius)
                                            {
                                                if (biome.leavesType[leavesType] != -1)
                                                    blocks[x + dx, dy, z + dz] =
                                                        biome.leavesType[leavesType];
                                            }
                                        }
                                    }
                                }

                                if (biome.leavesType[leavesType] != -1)
                                {
                                    blocks[x, height + treeHeight + 1, z] = biome.leavesType[leavesType];
                                    blocks[x + 1, height + treeHeight, z] = biome.leavesType[leavesType];
                                    blocks[x - 1, height + treeHeight, z] = biome.leavesType[leavesType];
                                    blocks[x, height + treeHeight, z + 1] = biome.leavesType[leavesType];
                                    blocks[x, height + treeHeight, z - 1] = biome.leavesType[leavesType];
                                }

                                for (int dy = 1; dy <= treeHeight; dy++)
                                    blocks[x, height + dy, z] = biome.treeType[treeType];
                            }
                        }
                    }
                }
            }
        }

        private void GenerateSurface(ref int[,,] blocks, out int[,] surfaceHeight, int xOffset, int yOffset,
            int zOffset)
        {
            surfaceHeight = new int[World.ChunkWidth, World.ChunkWidth];

            Biome biome = new Biome(_biomes[0]);

            for (int i = 0; i < World.ChunkWidth; i++)

            for (var x = 0; x < World.ChunkWidth; x++)
            {
                for (var z = 0; z < World.ChunkWidth; z++)
                {
                    var biomeType = (GetBiomeNoise(x + xOffset, z + zOffset) + 1f) / 2f;
                    for (int b = 0; b < _biomes.Length; b++)
                    {
                        if (biomeType <= _biomes[b].biomeChance)
                        {
                            biome = new Biome(_biomes[b]);
                            break;
                        }
                    }

                    var waterChance = _waterNoise.GetNoise(x + xOffset, z + zOffset);
                    var height = GetHeight(x + xOffset, z + zOffset);
                    if (waterChance > RiverChance)
                        height -= waterChance * WaterNoiseOctaves.Amplitude;

                    for (var y = 0; y <= (int)height && y < World.ChunkHeight; y++)
                    {
                        if (y < 1) blocks[x, y, z] = 1;
                        if (y >= 1 && y < height - 3) blocks[x, y, z] = 2;
                        if (y >= height - 3 && y < height - 1)
                            blocks[x, y, z] =
                                blocks[x, y, z] = (waterChance > RiverChance && height <= WaterLevel)
                                    ? 3
                                    : biome.subSurfaceBlock;
                        if (y >= height - 1 && y <= height)
                            blocks[x, y, z] = (waterChance > RiverChance && height <= WaterLevel)
                                ? 3
                                : biome.surfaceBlock;
                    }

                    surfaceHeight[x, z] = waterChance > RiverChance && height <= WaterLevel ? -1 : (int)height;
                }
            }
        }

        public int[,,] GenerateFlora(int xOffset, int yOffset, int zOffset)
        {
            var blocks = new int[World.ChunkWidth, World.ChunkHeight, World.ChunkWidth];
            
            Biome biome = new Biome(_biomes[0]);

            for (var x = 0; x < World.ChunkWidth; x++)
            {
                for (var z = 0; z < World.ChunkWidth; z++)
                {
                    var biomeType = (GetBiomeNoise(x + xOffset, z + zOffset) + 1f) / 2f;
                    for (int b = 0; b < _biomes.Length; b++)
                    {
                        if (biomeType <= _biomes[b].biomeChance)
                        {
                            biome = new Biome(_biomes[b]);
                            break;
                        }
                    }

                    var waterChance = _waterNoise.GetNoise(x + xOffset, z + zOffset);
                    var height = GetHeight(x + xOffset, z + zOffset);
                    if (waterChance > RiverChance)
                        height -= waterChance * WaterNoiseOctaves.Amplitude;
                    var floraChance = _floraNoise.GetNoise(x + xOffset, z + zOffset);
                    if (floraChance <= 0 && biome.placeFlora && (int)height > WaterLevel)
                    {
                        int floraIndex = 27;
                        var floraChances = biome.chances;
                        var fc = 0;
                        var currentFloraChance = Random.Range(0, floraChances);
                        var currentFloraChance2 = Random.Range(0, floraChances);
                        for (int i = 0; i < biome.floraId.Length; i++)
                        {
                            fc += biome.floraChances[i];
                            if (fc >= currentFloraChance)
                            {
                                floraIndex = biome.floraId[i];
                                break;
                            }
                        }

                        if (currentFloraChance2 <= biome.fullFloraChance)
                            blocks[x, (int)height + 1, z] = biome.placeFlora ? floraIndex : 0;
                    }
                }
            }

            return blocks;
        }

        public int[,,] GenerateWater(int xOffset, int yOffset, int zOffset)
        {
            var blocks = new int[World.ChunkWidth, World.ChunkHeight, World.ChunkWidth];
            for (var x = 0; x < World.ChunkWidth; x++)
            {
                for (var z = 0; z < World.ChunkWidth; z++)
                {
                    var waterChance = _waterNoise.GetNoise(x + xOffset, z + zOffset);
                    var height = GetHeight(x + xOffset, z + zOffset);
                    if (waterChance > RiverChance)
                        height -= waterChance * WaterNoiseOctaves.Amplitude;
                    for (var y = (int)height + 1; y <= WaterLevel; y++)
                    {
                        blocks[x, y, z] = -1;
                    }
                }
            }

            return blocks;
        }

        private float GetHeight(float x, float z)
        {
            float result = BaseHeight;

            for (int i = 0; i < _surfaceNoise.Length; i++)
            {
                float noise = _surfaceNoise[i].GetNoise(x, z);
                if (noise > 0)
                    result += noise * SurfaceNoiseOctaves[i].Amplitude / 2;
            }

            return result;
        }

        private float GetBiomeNoise(float x, float z)
        {
            _biomeNoise.DomainWarp(ref x, ref z);

            return _biomeNoise.GetNoise(x, z);
        }

        private float GetTreeNoise(float x, float z)
        {
            float result = 0;

            for (int i = 0; i < _treeNoise.Length; i++)
            {
                float noise = _treeNoise[i].GetNoise(x, z);
                result += noise * TreeNoiseOctaves[i].Amplitude / 2;
            }

            return result;
        }
    }
}

File: MultiCraft.Server/app/utils/WorldGenerator.js
import FastNoiseLite from 'fastnoise-lite';

export const worldGeneratorConfig = {
    BaseHeight: 64,
    WaterLevel: 63,
    RiverChance: -0.5,
    TreeFrequency: 0.5,
    Biomes: [
        {
            biomeName: "Desert",
            offset: 0,
            scale: 0,
            terrainHeight: 0,
            terrainScale: 0,
            surfaceBlock: 9,
            subSurfaceBlock: 9,
            treeType: [
                23
            ],
            leavesType: [
                -1
            ],
            maxTreeHeight: [
                3
            ],
            minTreeHeight: [
                1
            ],
            treeChance: [
                60
            ],
            biomeChance: 0.160999998,
            fullFloraChance: 0,
            floraIndex: 0,
            floraZoneScale: 0,
            floraThreshold: 0.119999997,
            floraPlacementScale: 0,
            floraPlacementThreshold: 0.100000001,
            placeFlora: false,
            floraId: [],
            floraChances: [],
            HeightMax: 0,
            HeightMin: 0,
            lodes: [],
            chances: 0
        },
        {
            biomeName: "Forest",
            offset: 0,
            scale: 0,
            terrainHeight: 0,
            terrainScale: 0,
            surfaceBlock: 4,
            subSurfaceBlock: 3,
            treeType: [
                5,
                98
            ],
            leavesType: [
                7,
                7
            ],
            maxTreeHeight: [
                7,
                7
            ],
            minTreeHeight: [
                5,
                5
            ],
            treeChance: [
                100,
                100
            ],
            biomeChance: 1,
            fullFloraChance: 100,
            floraIndex: 0,
            floraZoneScale: 0,
            floraThreshold: 0.100000001,
            floraPlacementScale: 0,
            floraPlacementThreshold: 0.100000001,
            placeFlora: true,
            floraId: [
                27,
                28,
                29
            ],
            floraChances: [
                80,
                10,
                10
            ],
            HeightMax: 0,
            HeightMin: 0,
            lodes: [],
            chances: 100
        },
        {
            biomeName: "ForestSpruce",
            offset: 0,
            scale: 0,
            terrainHeight: 0,
            terrainScale: 0,
            surfaceBlock: 1204,
            subSurfaceBlock: 3,
            treeType: [
                99
            ],
            leavesType: [
                578
            ],
            maxTreeHeight: [
                9
            ],
            minTreeHeight: [
                7
            ],
            treeChance: [
                75
            ],
            biomeChance: 0.226999998,
            fullFloraChance: 1,
            floraIndex: 0,
            floraZoneScale: 0,
            floraThreshold: 0.100000001,
            floraPlacementScale: 0,
            floraPlacementThreshold: 0.100000001,
            placeFlora: true,
            floraId: [
                556
            ],
            floraChances: [
                100
            ],
            HeightMax: 0,
            HeightMin: 0,
            lodes: [],
            chances: 0
        }
    ],
    BiomeNoiseOctaves: {
        NoiseType: "Cellular",
        Frequency: 0.00700000022,
        Amplitude: 2,
        FractalType: "DomainWarpProgressive",
        FractalOctaves: 6,
        FractalGain: 2.75,
        CellularDistanceFunction: "Euclidean",
        CellularReturnType: "CellValue",
        CellularJitter: 1,
        DomainWarpType: "OpenSimplex2Reduced",
        DomainWarpAmplitude: 4
    },
    FloraNoiseOctaves: {
        NoiseType: "Perlin",
        Frequency: 4.55999994,
        Amplitude: 1,
        FractalType: 0,
        FractalOctaves: 0,
        FractalGain: 0,
        CellularDistanceFunction: "Euclidean",
        CellularReturnType: "CellValue",
        CellularJitter: 0,
        DomainWarpType: "OpenSimplex2",
        DomainWarpAmplitude: 0
    },
    TreeNoiseOctaves: [
        {
            NoiseType: "Perlin",
            Frequency: 1.79999995,
            Amplitude: 2,
            FractalType: 0,
            FractalOctaves: 0,
            FractalGain: 0,
            CellularDistanceFunction: "Euclidean",
            CellularReturnType: "CellValue",
            CellularJitter: 0,
            DomainWarpType: "OpenSimplex2",
            DomainWarpAmplitude: 0
        }
    ],
    WaterNoiseOctaves: {
        NoiseType: "Cellular",
        Frequency: 0.00300000003,
        Amplitude: 4,
        FractalType: "Ridged",
        FractalOctaves: 1,
        FractalGain: 0.5,
        CellularDistanceFunction: "EuclideanSq",
        CellularReturnType: "Distance2Div",
        CellularJitter: 1,
        DomainWarpType: "OpenSimplex2",
        DomainWarpAmplitude: 12
    },
    SurfaceNoiseOctaves: [
        {
            NoiseType: "OpenSimplex2S",
            Frequency: 0.0120000001,
            Amplitude: 4,
            FractalType: 0,
            FractalOctaves: 0,
            FractalGain: 0,
            CellularDistanceFunction: "Euclidean",
            CellularReturnType: "CellValue",
            CellularJitter: 0,
            DomainWarpType: "OpenSimplex2",
            DomainWarpAmplitude: 0
        },
        {
            NoiseType: "OpenSimplex2S",
            Frequency: 0.0199999996,
            Amplitude: 12,
            FractalType: 0,
            FractalOctaves: 0,
            FractalGain: 0,
            CellularDistanceFunction: "Euclidean",
            CellularReturnType: "CellValue",
            CellularJitter: 0,
            DomainWarpType: "OpenSimplex2",
            DomainWarpAmplitude: 0
        },
        {
            NoiseType: "Perlin",
            Frequency: 1,
            Amplitude: 2,
            FractalType: 0,
            FractalOctaves: 0,
            FractalGain: 0,
            CellularDistanceFunction: "Euclidean",
            CellularReturnType: "CellValue",
            CellularJitter: 0,
            DomainWarpType: "OpenSimplex2",
            DomainWarpAmplitude: 0
        }
    ]
}
export class WorldGenerator {
    constructor(config, seed) {
        this.config = config;
        this.biomes = this.initializeBiomes();
        this.surfaceNoise = this.initializeNoiseGenerators(config.SurfaceNoiseOctaves, seed);
        this.biomeNoise = this.initializeNoiseGenerator(config.BiomeNoiseOctaves, seed);
        this.waterNoise = this.initializeNoiseGenerator(config.WaterNoiseOctaves, seed);
        this.floraNoise = this.initializeNoiseGenerator(config.FloraNoiseOctaves, seed);
        this.treeNoise = this.initializeNoiseGenerators(config.TreeNoiseOctaves, seed);
    }

    initializeBiomes() {
        const totalChance = this.config.Biomes.reduce((sum, biome) => sum + biome.biomeChance, 0);
        return this.config.Biomes.map(biome => ({
            ...biome,
            biomeChance: biome.biomeChance / totalChance // Normalize chances
        }));
    }

    initializeNoiseGenerator(NoiseOctave, seed) {
        const noise = new FastNoiseLite();
        noise.SetSeed(seed);
        noise.SetNoiseType(NoiseOctave.NoiseType);
        noise.SetFrequency(NoiseOctave.Frequency);
        noise.SetFractalType(NoiseOctave.FractalType);
        noise.SetFractalOctaves(NoiseOctave.FractalOctaves);
        noise.SetFractalGain(NoiseOctave.FractalGain);
        noise.SetCellularDistanceFunction(NoiseOctave.CellularDistanceFunction);
        noise.SetCellularReturnType(NoiseOctave.CellularReturnType);
        noise.SetCellularJitter(NoiseOctave.CellularJitter);
        noise.SetDomainWarpType(NoiseOctave.DomainWarpType);
        noise.SetDomainWarpAmp(NoiseOctave.DomainWarpAmplitude);
        return noise;
    }

    initializeNoiseGenerators(NoiseOctaves, seed) {
        return NoiseOctaves.map(octave => this.initializeNoiseGenerator(octave, seed));
    }

    generate(xOffset, zOffset) {
        const dimensions = { width: 16, height: 256, depth: 16 };
        const blocks = new Array(dimensions.width * dimensions.height * dimensions.depth).fill(0);
        const surfaceHeights = this.generateSurface(blocks, xOffset, zOffset);
        this.generateTrees(blocks, surfaceHeights, xOffset, zOffset);
        return blocks;
    }

    generateSurface(blocks, xOffset, zOffset) {
        const surfaceHeights = new Array(16 * 16).fill(0);

        for (let x = 0; x < 16; x++) {
            for (let z = 0; z < 16; z++) {
                const biome = this.getBiome(x + xOffset, z + zOffset);
                const waterChance = this.waterNoise.GetNoise(x + xOffset, z + zOffset);
                const height = Math.floor(this.getHeight(x + xOffset, z + zOffset) -
                    (waterChance > this.config.RiverChance ? waterChance * this.config.WaterNoiseOctaves.Amplitude : 0));

                this.fillColumn(blocks, x, z, height, biome, waterChance);
                surfaceHeights[x + z * 16] = height;
            }
        }

        return surfaceHeights;
    }

    fillColumn(blocks, x, z, height, biome, waterChance) {
        for (let y = 0; y < Math.min(height, 256); y++) {
            const blockType = this.getBlockType(y, height, biome, waterChance);
            if (blockType !== null) blocks[this.index3DTo1D(x, y, z)] = blockType;
        }
    }

    getBlockType(y, height, biome, waterChance) {
        if (y < 1) return 1; // Bedrock
        if (y < height - 3) return 2; // Stone
        if (y >= height - 3 && y < height - 1) {
            return waterChance > this.config.RiverChance && height <= this.config.WaterLevel ? 3 : biome.subSurfaceBlock;
        }
        if (y >= height - 1 && y <= height) {
            return waterChance > this.config.RiverChance && height <= this.config.WaterLevel ? 3 : biome.surfaceBlock;
        }
        return null;
    }

    generateTrees(blocks, surfaceHeights, xOffset, zOffset) {
        for (let x = 0; x < 16; x++) {
            for (let z = 0; z < 16; z++) {
                const height = surfaceHeights[x + z * 16];
                const treeNoiseValue = this.GetTreeNoise(x + xOffset, z + zOffset);

                if (treeNoiseValue > this.config.TreeFrequency && height > this.config.BaseHeight - 1) {
                    const biome = this.getBiome(x + xOffset, z + zOffset);
                    this.placeTree(blocks, x, z, height, biome);
                }
            }
        }
    }

    placeTree(blocks, x, z, height, biome) {
        if (biome.treeType.length === 0) return;

        const treeIndex = Math.floor(Math.random() * biome.treeType.length);
        const treeType = biome.treeType[treeIndex];
        const leavesType = biome.leavesType[treeIndex];
        const treeHeight = this.getRandomInRange(
            biome.minTreeHeight[treeIndex],
            biome.maxTreeHeight[treeIndex]
        );

        if (Math.random() * 100 > biome.treeChance[treeIndex]) return;

        this.generateTreeBlocks(blocks, x, z, height, treeHeight, treeType, leavesType);
    }

    generateTreeBlocks(blocks, x, z, height, treeHeight, treeType, leavesType) {
        // Generate foliage
        for (let y = height + treeHeight - 1; y >= height + treeHeight / 2; y--) {
            const foliageRadius = Math.max(1, Math.min(3, treeHeight - (height + treeHeight - y)));
            this.generateFoliageLayer(blocks, x, z, y, foliageRadius, leavesType);
        }

        // Generate trunk
        for (let y = 0; y < treeHeight; y++) {
            blocks[this.index3DTo1D(x, height + y, z)] = treeType;
        }
    }

    generateFoliageLayer(blocks, x, z, y, radius, leavesType) {
        for (let dx = -radius; dx <= radius; dx++) {
            for (let dz = -radius; dz <= radius; dz++) {
                if (dx * dx + dz * dz <= radius * radius) {
                    blocks[this.index3DTo1D(x + dx, y, z + dz)] = leavesType;
                }
            }
        }
    }

    getBiome(x, z) {
        const biomeNoiseValue = (this.GetBiomeNoise(x, z) + 1) / 2;
        let accumulatedChance = 0;

        for (const biome of this.biomes) {
            accumulatedChance += biome.biomeChance;
            if (biomeNoiseValue <= accumulatedChance) return biome;
        }

        return this.biomes[this.biomes.length - 1];
    }

    getHeight(x, z) {
        return this.config.BaseHeight + this.surfaceNoise.reduce((height, noise, i) =>
            height + noise.GetNoise(x, z) * this.config.SurfaceNoiseOctaves[i].Amplitude / 2, 0);
    }

    GetTreeNoise(x, z) {
        return this.treeNoise.reduce((total, noise, i) =>
            total + noise.GetNoise(x, z) * this.config.TreeNoiseOctaves[i].Amplitude / 2, 0);
    }

    GetBiomeNoise(x, z) {
        return this.biomeNoise.GetNoise(x, z);
    }

    index3DTo1D(x, y, z) {
        return x + z * 16 + y * 256;
    }

    getRandomInRange(min, max) {
        return Math.floor(Math.random() * (max - min + 1)) + min;
    }
}